toolchain/check/eval.cpp

```diff
@@ -553,6 +573,62 @@ static auto PerformBuiltinIntComparison(Context& context,
   return MakeBoolResult(context, bool_type_id, result);
 }
 
+// Performs a builtin unary float -> float operation.
+static auto PerformBuiltinUnaryFloatOp(Context& context,
+                                       SemIR::BuiltinFunctionKind builtin_kind,
+                                       SemIR::InstId arg_id)
+    -> SemIR::ConstantId {
+  auto op = context.insts().GetAs<SemIR::FloatLiteral>(arg_id);
+  auto op_val = context.floats().Get(op.float_id);
+
+  switch (builtin_kind) {
+    case SemIR::BuiltinFunctionKind::FloatNegate:
+      op_val = llvm::APFloat(0.0) - op_val;
+      break;
+    default:
+      CARBON_FATAL() << "Unexpected builtin kind";
+  }
+
+  return MakeFloatResult(context, op.type_id, std::move(op_val));
+}
+
+// Performs a builtin binary float -> float operation.
+static auto PerformBuiltinBinaryFloatOp(Context& context, SemIRLoc loc,
+                                        SemIR::BuiltinFunctionKind builtin_kind,
+                                        SemIR::InstId lhs_id,
+                                        SemIR::InstId rhs_id)
+    -> SemIR::ConstantId {
+  auto lhs = context.insts().GetAs<SemIR::FloatLiteral>(lhs_id);
+  auto rhs = context.insts().GetAs<SemIR::FloatLiteral>(rhs_id);
+  auto lhs_val = context.floats().Get(lhs.float_id);
+  auto rhs_val = context.floats().Get(rhs.float_id);
+
+  llvm::APFloat result_val(0.0);
+
+  switch (builtin_kind) {
+    case SemIR::BuiltinFunctionKind::FloatAdd:
+      result_val = lhs_val + rhs_val;
+      break;
+    case SemIR::BuiltinFunctionKind::FloatSub:
+      result_val = lhs_val - rhs_val;
+      break;
+    case SemIR::BuiltinFunctionKind::FloatMul:
+      result_val = lhs_val * rhs_val;
+      break;
+    case SemIR::BuiltinFunctionKind::FloatDiv:
+      if (rhs_val.isZero()) {
+        DiagnoseDivisionByZero(context, loc);
+        return SemIR::ConstantId::Error;
+      }
```

Comment:
We can drop this check; per [the design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/arithmetic.md#:~:text=division%20by%20zero%20results%20in%20either%20%C2%B1%E2%88%9E%20or%2C%20for%200.0%20/%200.0%2C%20a%20quiet%20NaN.), floating-point division by zero is permitted.

Reply:
Done!